File: web/src/store/rss.ts
// RSS store: the /api/rss view (feeds, items, filters) plus refresh and
// item actions (PAR-3.3). The sidebar unread badge and the RSS view share it.

import { createSignal } from "solid-js";
import { showToast } from "./ui";

export interface RssFeed {
  name: string;
  url: string;
  label: string;
  destination: string;
  pollIntervalNs: number;
  lastFetch?: string;
  lastOk?: string;
  lastError?: string;
  retryInSecs: number;
  items: number;
  unread: number;
}

export interface RssItem {
  feed: string;
  id: string;
  title: string;
  link: string;
  enclosureUrl: string;
  enclosureType?: string;
  length: number;
  categories: string[];
  published?: string;
  read: boolean;
  loaded: boolean;
  loadedHash?: string;
  loadedBy?: string;
}

export interface RssEval {
  at: string;
  feed: string;
  itemId: string;
  title: string;
  outcome: string;
  reason: string;
}

export interface RssFilter {
  name: string;
  feed?: string;
  evaluated: number;
  matched: number;
  loaded: number;
  history: RssEval[];
}

export interface RssView {
  feeds: RssFeed[];
  items: RssItem[];
  filters: RssFilter[];
}

const [view, setView] = createSignal<RssView>({ feeds: [], items: [], filters: [] });
const [loading, setLoading] = createSignal(false);
const [failed, setFailed] = createSignal(false);

export { view, loading, failed };

/** Total unread, unloaded items across feeds, for the sidebar badge. */
export function unreadCount(): number {
  return view().feeds.reduce((sum, feed) => sum + (feed.unread || 0), 0);
}

export async function refreshRss(quiet = false): Promise<void> {
  setLoading(true);
  setFailed(false);
  try {
    const response = await fetch("/api/v1/rss", { headers: { Accept: "application/json" } });
    if (!response.ok) throw new Error("Could not load RSS feeds");
    setView((await response.json()) as RssView);
  } catch (error) {
    setFailed(true);
    if (!quiet) showToast(error instanceof Error ? error.message : "Could not load RSS feeds");
  } finally {
    setLoading(false);
  }
}

export async function addRssItem(feed: string, id: string): Promise<void> {
  try {
    const response = await fetch("/api/v1/rss/add", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ feed, id }),
    });
    const body = (await response.json().catch(() => ({}))) as {
      hash?: string;
      error?: { message?: string };
    };
    if (!response.ok) throw new Error(body.error?.message || "Could not add torrent");
    showToast("Torrent added to session.");
    await refreshRss();
  } catch (error) {
    showToast(error instanceof Error ? error.message : "Could not add torrent");
  }
}

export async function markRssRead(feed?: string): Promise<void> {
  try {
    const response = await fetch("/api/v1/rss/read", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(feed ? { feed, all: true } : { all: true }),
    });
    if (!response.ok) throw new Error("Could not mark items read");
    await refreshRss();
  } catch (error) {
    showToast(error instanceof Error ? error.message : "Could not mark items read");
  }
}

File: internal/api/rss.go
package api

import (
	"encoding/json"
	"net/http"
	"time"

	"blackbird/internal/config"
	"blackbird/internal/rss"
)

// secretMask stands in for stored feed credentials in the settings API. A
// submitted mask value keeps the stored secret; an empty value clears it.
const secretMask = "***"

// ---- GET /api/rss, POST /api/rss/add, POST /api/rss/read (PAR-3.3) ----

type rssFeedDTO struct {
	Name         string `json:"name"`
	URL          string `json:"url"`
	Label        string `json:"label"`
	Destination  string `json:"destination"`
	PollInterval int64  `json:"pollIntervalNs"`
	LastFetch    string `json:"lastFetch,omitempty"`
	LastOK       string `json:"lastOk,omitempty"`
	LastError    string `json:"lastError,omitempty"`
	RetryInSecs  int64  `json:"retryInSecs"`
	Items        int    `json:"items"`
	Unread       int    `json:"unread"`
}

type rssItemDTO struct {
	Feed          string   `json:"feed"`
	ID            string   `json:"id"`
	Title         string   `json:"title"`
	Link          string   `json:"link"`
	EnclosureURL  string   `json:"enclosureUrl"`
	EnclosureType string   `json:"enclosureType,omitempty"`
	Length        int64    `json:"length"`
	Categories    []string `json:"categories"`
	Published     string   `json:"published,omitempty"`
	Read          bool     `json:"read"`
	Loaded        bool     `json:"loaded"`
	LoadedHash    string   `json:"loadedHash,omitempty"`
	LoadedBy      string   `json:"loadedBy,omitempty"`
}

type rssEvalDTO struct {
	At      string `json:"at"`
	Feed    string `json:"feed"`
	ItemID  string `json:"itemId"`
	Title   string `json:"title"`
	Outcome string `json:"outcome"`
	Reason  string `json:"reason"`
}

type rssFilterDTO struct {
	Name      string       `json:"name"`
	Feed      string       `json:"feed,omitempty"`
	Evaluated int          `json:"evaluated"`
	Matched   int          `json:"matched"`
	Loaded    int          `json:"loaded"`
	History   []rssEvalDTO `json:"history"`
}

type rssViewDTO struct {
	Feeds   []rssFeedDTO   `json:"feeds"`
	Items   []rssItemDTO   `json:"items"`
	Filters []rssFilterDTO `json:"filters"`
}

func formatRSSTime(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	return t.UTC().Format(time.RFC3339)
}

func (s *Server) rssService() *rss.Service {
	if s.opts.RSS == nil {
		return nil
	}
	return s.opts.RSS
}

func (s *Server) rssViewHandler(w http.ResponseWriter, r *http.Request) {
	svc := s.rssService()
	if svc == nil {
		writeAPIError(w, http.StatusServiceUnavailable, "unavailable", "rss service is not running")
		return
	}
	view := svc.Snapshot()
	out := rssViewDTO{Feeds: []rssFeedDTO{}, Items: []rssItemDTO{}, Filters: []rssFilterDTO{}}
	for _, f := range view.Feeds {
		out.Feeds = append(out.Feeds, rssFeedDTO{
			Name: f.Name, URL: f.URL, Label: f.Label, Destination: f.Destination,
			PollInterval: int64(f.PollInterval), LastFetch: formatRSSTime(f.LastFetch),
			LastOK: formatRSSTime(f.LastOK), LastError: f.LastError,
			RetryInSecs: int64(f.RetryIn / time.Second), Items: f.Items, Unread: f.Unread,
		})
	}
	for _, item := range view.Items {
		out.Items = append(out.Items, rssItemDTO{
			Feed: item.Feed, ID: item.ID, Title: item.Title, Link: item.Link,
			EnclosureURL: item.EnclosureURL, EnclosureType: item.EnclosureType,
			Length: item.Length, Categories: item.Categories,
			Published: formatRSSTime(item.Published), Read: item.Read,
			Loaded: item.Loaded, LoadedHash: item.LoadedHash, LoadedBy: item.LoadedBy,
		})
	}
	for _, f := range view.Filters {
		dto := rssFilterDTO{
			Name: f.Name, Feed: f.Feed, Evaluated: f.Evaluated,
			Matched: f.Matched, Loaded: f.Loaded, History: []rssEvalDTO{},
		}
		for _, e := range f.History {
			dto.History = append(dto.History, rssEvalDTO{
				At: formatRSSTime(e.At), Feed: e.Feed, ItemID: e.ItemID,
				Title: e.Title, Outcome: e.Outcome, Reason: e.Reason,
			})
		}
		out.Filters = append(out.Filters, dto)
	}
	writeJSON(w, http.StatusOK, out)
}

type rssAddRequest struct {
	Feed string `json:"feed"`
	ID   string `json:"id"`
}

func (s *Server) rssAddHandler(w http.ResponseWriter, r *http.Request) {
	svc := s.rssService()
	if svc == nil {
		writeAPIError(w, http.StatusServiceUnavailable, "unavailable", "rss service is not running")
		return
	}
	var req rssAddRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeAPIError(w, http.StatusBadRequest, "bad_request", "invalid JSON body: "+err.Error())
		return
	}
	if req.Feed == "" || req.ID == "" {
		writeAPIError(w, http.StatusBadRequest, "bad_request", "feed and id are required")
		return
	}
	hash, err := svc.AddItem(req.Feed, req.ID)
	if err != nil {
		writeAPIError(w, http.StatusBadRequest, "rss_add_failed", err.Error())
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"hash": hash})
}

type rssReadRequest struct {
	Feed string   `json:"feed,omitempty"`
	IDs  []string `json:"ids,omitempty"`
	All  bool     `json:"all,omitempty"`
}

func (s *Server) rssReadHandler(w http.ResponseWriter, r *http.Request) {
	svc := s.rssService()
	if svc == nil {
		writeAPIError(w, http.StatusServiceUnavailable, "unavailable", "rss service is not running")
		return
	}
	var req rssReadRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeAPIError(w, http.StatusBadRequest, "bad_request", "invalid JSON body: "+err.Error())
		return
	}
	marked := svc.MarkRead(req.Feed, req.IDs, req.All || len(req.IDs) == 0)
	writeJSON(w, http.StatusOK, map[string]any{"marked": marked})
}

// ---- Secret redaction for the settings API ----

// redactAutomation returns a copy of the automation section with feed
// credentials replaced by the mask, so secrets never leave the server.
func redactAutomation(a config.Automation) config.Automation {
	out := a
	out.Rss.Feeds = make([]config.RSSFeed, len(a.Rss.Feeds))
	for i, f := range a.Rss.Feeds {
		out.Rss.Feeds[i] = f
		if f.Cookies != "" {
			out.Rss.Feeds[i].Cookies = secretMask
		}
		if len(f.Headers) > 0 {
			masked := make(map[string]string, len(f.Headers))
			for k := range f.Headers {
				masked[k] = secretMask
			}
			out.Rss.Feeds[i].Headers = masked
		}
	}
	return out
}

// restoreRSSSecrets merges submitted automation rules with the stored config:
// mask values keep the stored secret, empty values clear it. Feeds are
// matched by name.
func restoreRSSSecrets(changed *config.Automation, current config.Automation) {
	stored := map[string]config.RSSFeed{}
	for _, f := range current.Rss.Feeds {
		stored[f.Name] = f
	}
	for i := range changed.Rss.Feeds {
		sub := &changed.Rss.Feeds[i]
		prev, ok := stored[sub.Name]
		if !ok {
			// New feed: a mask value with no stored secret means nothing.
			if sub.Cookies == secretMask {
				sub.Cookies = ""
			}
			for k, v := range sub.Headers {
				if v == secretMask {
					delete(sub.Headers, k)
				}
			}
			if len(sub.Headers) == 0 {
				sub.Headers = nil
			}
			continue
		}
		if sub.Cookies == secretMask {
			sub.Cookies = prev.Cookies
		}
		if sub.Headers == nil {
			continue
		}
		merged := map[string]string{}
		for k, v := range sub.Headers {
			if v == secretMask {
				if kept, ok := prev.Headers[k]; ok {
					merged[k] = kept
				}
				continue
			}
			merged[k] = v
		}
		sub.Headers = merged
		if len(sub.Headers) == 0 {
			sub.Headers = nil
		}
	}
}
